### PR TITLE
Ewm 4058 - Automate State Initialization

### PR DESCRIPTION
**Pull request by darshdinger**
## Description of work
This task focused on refining the initialization state workflow, introducing an automated verification feature that activates each time a run number is submitted. Now, SNAPRed efficiently determines if the provided run number is linked to an existing state. If a match is found, the process seamlessly continues as planned. Conversely, if no associated state is detected, the system initiates an alternative workflow, offering users the option to initialize the state, thereby ensuring a smoother and more intuitive user experience.

## Explanation of work
The introduction of this feature presents a unique challenge in terms of complexity, given the intricate and somewhat enigmatic aspects of SNAPRed's operational flow. However, I will endeavor to demystify how this enhancement works.

Embedded within the Calibration and Normalization workflows are input fields where users can enter a run number for assessment. After completing the required fields, users proceed by clicking the Continue button. The logic governing this action is encapsulated within the WorkflowPresenter, specifically in the _handleContinueButtonClicked method. This method now includes a refined process, _handleComplications, engineered to accommodate a newly defined error type, RecoverableError.

On the backend, particularly within the error management directory, we've introduced a bespoke exception named RecoverableError. This is strategically employed alongside the ExecutionHandler in the LocalDataService to intercept exceptions triggered by the readCalibrationState method. Furthermore, adjustments to SNAPResponse have been made to facilitate the detection of such exceptions, thereby triggering the initiation of the state initialization workflow. This comprehensive overhaul not only integrates this advanced feature into the existing framework but also enhances the interface and workflow mechanisms for a smoother user interaction.

## To test

To test this feature, open SNAPRed GUI and complete the DiffractionCalibration Workflow form. Values used:
```python
Run Number: 46581
Convergence Threshold: 0.1
Peak Intensity Threshold: 0.0001
Bins Across Peak Width: 10
Sample: Any
Grouping File: Any
Peak Function: Any
``` 

This should result in a pop-up that prompts the user if they want to initialize the state. Please test this feature to completion. Test with other non-initialized run numbers.

## Link to EWM item

[EWM# 4058](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=4058)

### CODE REVIEW

File: src/snapred/backend/error/StateValidationException.py [hunk 1]
```diff
@@ -7,6 +9,16 @@ class StateValidationException(Exception):
     "Raised when an Instrument State is invalid"
 
     def __init__(self, exception: Exception):
-        self.message = "Instrument State for given Run Number is invalid! (see logs for details.)"
-        logger.error(str(exception))
+        exceptionStr = str(exception)
+
+        path_pattern = r"(/[^'\"]+)"
```
Comment (rboston628) [inline src/snapred/backend/error/StateValidationException.py:14]:
```
(>[^'_'^])>oc|==> 
```

File: src/snapred/backend/error/RecoverableException.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+from typing import Any, Literal
+
+from snapred.backend.log.logger import snapredLogger
+
+logger = snapredLogger.getLogger(__name__)
+
+RecoverableErrorType = Literal[
+    "AttributeError: 'NoneType' object has no attribute 'instrumentState'",
+    "etc",
+]
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/backend/error/RecoverableException.py:10]: Michael told me to make a generic exception type for later use. This is the reason for this.

File: src/snapred/backend/api/InterfaceController.py [hunk 3]
```diff
@@ -23,22 +24,26 @@ def executeRequest(self, request: SNAPRequest) -> SNAPResponse:
...
 
             message = None
             if snapredLogger.hasWarnings():
                 message = self.getWarnings()
-                snapredLogger.clearWarnings()
             response = SNAPResponse(code=200, message=message, data=result)
+
+        except RecoverableException as e:
+            self.logger.error(f"Recoverable error occurred: {str(e)}")
+            response = SNAPResponse(code=400, message="state")
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/backend/api/InterfaceController.py:38]: This is again per Michael's request, he said to initialize this error with a message of 'state' in order to differentiate it from other errors we may handle using codes within the 400's later on.

File: src/snapred/backend/error/StateValidationException.py (hunk 1 above)
Comment (darshdinger) [reply src/snapred/backend/error/StateValidationException.py:14]: Hate regex.

File: src/snapred/backend/error/StateValidationException.py [hunk 4]
```diff
@@ -7,6 +9,16 @@ class StateValidationException(Exception):
...
+        match = re.search(path_pattern, exceptionStr)
+
+        if "/SNS/SNAP/" in exceptionStr and match:
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/backend/error/StateValidationException.py:17]: This is reading a path type so there the only instance I can see this happening is if SNAPRed is used outside of analysis on a local based directory. Maybe I should change this to use the application.yml file.

File: src/snapred/backend/error/RecoverableException.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+
+class RecoverableException(Exception):
+    """
+    Raised when a recoverable exception is thrown.
+    Allows for custom error types and optional extra context.
+    """
+
+    def __init__(self, exception: Exception, errorType: RecoverableErrorType, **kwargs: Any):
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/backend/error/RecoverableException.py:19]: Switched to `errorMsg`.

'calibrationState is None' because from within `LocalDataService` the method used to check state checks based on a path that is constructed with a set of hex values (stateID). The returned variable remains `None` if this path does not exist (e.i. os.path.exists(calibrationStatePath)). It does not return the error ['"AttributeError: 'NoneType' object has no attribute 'instrumentState'"] until later on in the execution.

File: src/snapred/backend/data/LocalDataService.py [hunk 6]
```diff
@@ -614,6 +615,7 @@ def getNormalizationStatePath(self, runId: str, version: str):
         )
         return statePath
 
+    @ExceptionHandler(RecoverableException, "'NoneType' object has no attribute 'instrumentState'")
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/backend/data/LocalDataService.py:618]: Could be, but I will leave for now.

File: src/snapred/backend/error/RecoverableException.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        self.errorType = errorType
+        if errorType:
+            self.message = errorType
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/backend/error/RecoverableException.py:22]: This is true.

[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/backend/error/RecoverableException.py:22]: Fixed

File: src/snapred/ui/presenter/InitializeStatePresenter.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+from PyQt5.QtWidgets import QMessageBox
+
+from snapred.backend.api.InterfaceController import InterfaceController
+from snapred.backend.dao.request.InitializeStateRequest import InitializeStateRequest
+from snapred.backend.dao.SNAPRequest import SNAPRequest
+from snapred.backend.dao.SNAPResponse import SNAPResponse
+from snapred.ui.threading.worker_pool import WorkerPool
+
+
+class InitializationWorkflow(QObject):
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/ui/presenter/InitializeStatePresenter.py:11]: It could but I felt better to put this within a Presenter/Workflow type file as it has a view associated with it.

File: src/snapred/ui/presenter/InitializeStatePresenter.py
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/ui/presenter/InitializeStatePresenter.py:1]: Fixed

File: src/snapred/ui/presenter/WorkflowPresenter.py [hunk 9]
```diff
@@ -127,15 +127,35 @@ def handleContinueButtonClicked(self, model):
 
         self.worker_pool.submitWorker(self.worker)
 
+    def _isErrorCode(self, code):
+        return code - 499 > 0
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/ui/presenter/WorkflowPresenter.py:131]: OKAY! Done

File: src/snapred/ui/presenter/InitializeStatePresenter.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    def __init__(self, view):
+        super().__init__()
+        self.view = view
+        self.interfaceController = InterfaceController()
+
+    def handleButtonClicked(self):
+        runNumber = self.view.getRunNumber()
+        stateName = self.view.getStateName()
+
+        if not runNumber.isdigit():
```
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/ui/presenter/InitializeStatePresenter.py:24]: Done

File: src/snapred/backend/api/InterfaceController.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rboston628) [reply src/snapred/backend/api/InterfaceController.py:38]: can you change it to
``` python
response = SNAPResponse(code=ResponseCode.RECOVERABLE, message="state")
```

File: src/snapred/backend/error/StateValidationException.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rboston628) [reply src/snapred/backend/error/StateValidationException.py:17]:
``` python
if Config["instrument.home"] in exceptionStr:
```

File: src/snapred/backend/api/InterfaceController.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (darshdinger) [reply src/snapred/backend/api/InterfaceController.py:38]: I tried this, seemed to cause the tests to fail and say that ResponseCode.RECOVERABLE doesn't exist

**Comment by rboston628** [comment]
# Testing, Captain's Log

## DiffCal Workflow :white_check_mark: 

I open the test panel, and the groupings are pre-populated with default groupings.  I type in 46680 (initialized state) and the groupings are populated with the ones for thus state.  I then type in 46581 as directed (no state) and groupings correctly load only the defaults.  This shows this work did not revert a previous PR on the UX.
image

I enter run 46581, convergence threshold 0.1, peak threshold 0.0001, bins 10, and choose silicon, column, and grouping.  I click continue, and am greeted by this popup:
image

I select yes, enter the run number (would be good if it pre-populated), and success!
image

**Clicking the button does not close the "Initialization Menu" window, which is still there and has to be closed manually.**

I manually close the initialization window, the click to continue.  Wheels turn, and then this window appears, showing some of the most awful data I've ever seen:
image

Of course, this probably isn't a silicon data run, so the predicted and actual peaks don't match.  This ugliness of the focused data here should be looked at in a future PR.
image

However, this shows that this did not revert the PR to show the peak-tweak view in diffcal workflow.

I continue, the run completes.  I go to save, save it, and this friendly pop up occurs.
image

I am then returned to the starting screen with all of my data erased.  I strongly disagree with that behavior, but I think either Michael or Malcolm requested it and it isn't part of this PR, so I will say no more.

## Normalization Workflow :x:

I delete the state folder, reopen, try normalization.

Firstly,** this was re-named back to "Normalization Calibration" which makes the vertical tabs go off the screen.  Rename this tab to "Normalization" and possibly resize the window to fit all the tabs vertically.**

With no run number it shows the default groupings.  With an invalid number is shows default.  With 46581 it shows the default groupings.  With 46680 it shows it shows the state-specific groupings.  So previous work on this was not reverted by this PR.

I try 46581 and enter background 58813 (a different state), select diamond and bank groupings.  A helpful error message from previous work pops up informing me I can't have a background from a different state.  Good.

I enter 46581 twice and press to continue.  Same popup asking to initialize it
image
I select yes, initialize it, enter the run number (**this should prepopulate**) it finishes, and I can press okay.  

**The initialization window is still there and needs to go away**

I click to continue.  It calculates the raw vanadium correction, but then this happens:
image

It might be hard to see but the window has frozen before resizing.  It then exits with a segmentation fault.
image
In the UI, this is usually caused by threading errors.

If I re-open SNAPRed with the state for 46581 already initialized, select the same runs and files then select to continue, I get to the peak-tweak view
image

I can then continue and get to saving, which completes successfully and deletes all my stuff.
image

The initial segfault needs to be fixed

## TLDR

Needs some minor fixes.
- [ ] initialization menu window should pre-populate the run number field
- [ ] clicking "Okay" on the pop-up for a successful initialization should close the initialization menu window as well
- [ ] the "process completed" successfully popup only needs one button just saying "Okay", since both options only close the window
- [ ] **The segfault when running with normalization needs to be fixed**

File: src/snapred/backend/api/InterfaceController.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rboston628) [reply src/snapred/backend/api/InterfaceController.py:44]: did you import ResponseCode?

**Review by dlcaballero16 — APPROVED** [review]
Works as expected when testing on analysis.
